blog/2018/12/20/testing-dancer/index.markdown

```diff
@@ -0,0 +1,241 @@
+If you missed it, go [check it out](https://mojolicious.io/blog/2017/12/09/day-9-the-best-way-to-test/).
+
+I expect there are at least a few of you out there who read that and think, "I'd love to use that, but I don't use Mojolicious!"; well, you're in luck!
+With just a little role to bridge the gap, you can use Test::Mojo to test your PSGI applications too!
+
+---
+
+## Mounting PSGI Applications
+
+Mojolicious itself doesn't use the [PSGI](https://metacpan.org/pod/PSGI) protocol, owing to certain features that it doesn't provide and which are necessary for certain asynchronous operations.
+That said, you can serve a Mojolicious application on a PSGI server by using [Mojo::Server::PSGI](https://mojolicious.org/perldoc/Mojo/Server/PSGI).
+This Mojolicious-core module is automatically used for you when your Mojolicious-based app detects that it has started under a PSGI server (e.g. plackup or Starman).
+
+While translating between a Mojo app and a PSGI server is core functionality, doing the opposite, translating between a PSGI app and a Mojolicious server (or app, as you'll see) is available as a third party module.
+[Mojolicious::Plugin::MountPSGI](https://metacpan.org/pod/Mojolicious::Plugin::MountPSGI), as it's name implies, can mount a PSGI application into a Mojolicious-based one.
+To do so, it builds a new, empty Mojolicious application that translates all requests to PSGI environments before dispatching to it as with any [mount](https://mojolicious.org/perldoc/Mojolicious/Plugin/Mount)-ed application.
+
+## Testing using Test::Mojo
+
+Once you can do that, it is trivial to take a PSGI application, wrap it with MountPSGI, and set it as the application for use with Test::Mojo.
+Still, to make it even easier, that has all been done for you in [Test::Mojo::Role::PSGI](https://metacpan.org/pod/Test::Mojo::Role::PSGI).
+
+Like any [Mojolicious Role](https://mojolicious.io/blog/2017/12/13/day-13-more-about-roles/), we can use `with_roles` to create a (mostly anonymous) subclass with the role applied.
+You can use the shortcut `+` to stand in for `Test::Mojo::Role::`.
+
+    use Test::Mojo;
+    my $class = Test::Mojo->with_roles('+PSGI');
+
+Then you instantiate that role with the path to the PSGI application, or else the PSGI application itself.
+
+Since you're using roles, which are all about composition, you can also apply other roles that you might [find on CPAN](https://metacpan.org/search?q=%22Test%3A%3AMojo%3A%3ARole%22).
+
+## An Example
+
+As an example, let's say we have a simple application script (named `app.psgi`) that can render a `"hello world"` or `"hello $user"` in several formats.
+I'll allow a plain text response, JSON, and templated HTML (using the [simple](https://metacpan.org/pod/Dancer2::Template::Simple) template to keep this concise).
+
+    use Dancer2;
+
+    set template => 'simple';
+    set views => '.';
+
+    any '/text' => sub {
+      my $name = param('name') // 'world';
```

Comment:
You should replace `param()` with `body_parameters->get()`. While `param()` still works, `body_parameters` does so in a safer way.

Reply:
In the tests I actually switch between get and post so it was kinda handy to not care between body or query parameters. I suppose I don't have to and I don't even comment on it directly. Is there a "safer" version of "any urlencoded parameter"?

Reply:
I missed the GET, sorry.

Sadly, there isn't (though you're giving me ideas). So, given your particular case, I'd stick with `param()` for now.